### PR TITLE
Fix only work package custom_options being avaialable in API

### PR DESCRIPTION
**Pull request by oliverguenther**
Drive-by fix while confirming bug

https://community.openproject.org/wp/44281

### CODE REVIEW

File: lib/api/v3/custom_options/custom_options_api.rb [hunk 1]
```diff
@@ -37,12 +37,26 @@ class CustomOptionsAPI < ::API::OpenProjectAPI
             end
 
             helpers do
-              def authorize_view_in_activated_project(custom_option)
+              def authorize_custom_option_visibility(custom_option)
+                case custom_option.custom_field
+                when WorkPackageCustomField
+                  authorized_work_package_option(custom_option)
+                when ProjectCustomField
+                  authorize_any(%i[view_project], global: true) { raise API::Errors::NotFound.new }
```
Comment (dombesz) [inline lib/api/v3/custom_options/custom_options_api.rb]: I think we can omit the `.new` and just use `raise API::Errors::NotFound` as below.

File: app/models/custom_option.rb [hunk 2]
```diff
@@ -38,6 +38,8 @@ class CustomOption < ApplicationRecord
 
   before_destroy :assure_at_least_one_option
 
+  delegate :field_format, to: :custom_field
```
Comment (dombesz) [inline app/models/custom_option.rb]: Are we using this delegate?

**Review by dombesz — COMMENTED** [review]
Not sure if this is the intended behaviour, but when I remove all the permissions for a user, I am still able to view the custom options for a `WorkPackageCustomField`. Inspecting the query further, the query for `:view_work_packages` permission does not include a permission check.

```ruby
[4] pry(#<#<Class:0x000000011d80d478>>)> Authorization.projects(:view_work_packages, current_user)
  Project Load (3.0ms)  SELECT "projects".* FROM "projects" INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id" AND "enabled_modules"."name" IN ('work_package_tracking') AND "projects"."active" = TRUE WHERE "projects"."active" = TRUE
```

(There are some failing tests too)

File: lib/api/v3/custom_options/custom_options_api.rb (hunk 1 above)
Comment (oliverguenther) [reply lib/api/v3/custom_options/custom_options_api.rb]: Yes, will do

**Comment by oliverguenther** [comment]
> Not sure if this is the intended behaviour, but when I remove all the permissions for a user, I am still able to view the custom options for a `WorkPackageCustomField`. Inspecting the query further, the query for `:view_work_packages` permission does not include a permission check.
> 
> ```ruby
> [4] pry(#<#<Class:0x000000011d80d478>>)> Authorization.projects(:view_work_packages, current_user)
>   Project Load (3.0ms)  SELECT "projects".* FROM "projects" INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id" AND "enabled_modules"."name" IN ('work_package_tracking') AND "projects"."active" = TRUE WHERE "projects"."active" = TRUE
> ```
> 
> (There are some failing tests too)

You're likely checking with an admin user, which short circuits to all projects that have work package tracking active. Here's a regular user's resulting query:

```sql
SELECT "projects".* FROM "projects" LEFT OUTER JOIN "members" ON "projects"."id" = "members"."project_id" AND "members"."user_id" = 73049 AND "projects"."active" = TRUE INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id" AND "enabled_modules"."name" IN ('work_package_tracking') AND "projects"."active" = TRUE INNER JOIN "role_permissions" ON "role_permissions"."permission" IN ('view_work_packages') INNER JOIN "roles" "permission_roles" ON "permission_roles"."id" = "role_permissions"."role_id" LEFT OUTER JOIN "member_roles" ON "members"."id" = "member_roles"."member_id" LEFT OUTER JOIN "roles" "assigned_roles" ON "assigned_roles"."id" = "permission_roles"."id" AND "projects"."active" = TRUE AND ("assigned_roles"."id" = "member_roles"."role_id" OR "projects"."public" = TRUE AND "assigned_roles"."builtin" = 1 AND "member_roles"."id" IS NULL) WHERE "assigned_roles"."id" IS NOT NULL
```

**Review by dombesz — APPROVED** [review]
LGTM! 👍

**Pull request merged**